src/MentorBot.Data.Common/DbRepository{T}.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) 2018. Licensed under the MIT License. See https://www.opensource.org/licenses/mit-license.php for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+using MentorBot.Data.Common.Models;
+
+using Microsoft.EntityFrameworkCore;
+
+namespace MentorBot.Data.Common
+{
+    /// <summary>
+    /// The generic entity framework repository.
+    /// </summary>
+    /// <typeparam name="T">The exact model type.</typeparam>
+    /// <seealso cref="MentorBot.Data.Common.IDbRepository{T}" />
+    /// <seealso cref="System.IDisposable" />
+    public class DbRepository<T> : IDbRepository<T>, IDisposable
+        where T : class, IAuditInfo, IDeletableEntity
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DbRepository{T}"/> class.
+        /// </summary>
+        public DbRepository(DbContext dbContext)
+        {
+            var context = dbContext ?? throw new ArgumentException("An instance of DbContext is required to use this repository.", nameof(dbContext));
+
+            this.Context = context;
+            this.DbSet = this.Context.Set<T>();
+        }
+
+        private DbContext Context { get; }
+
+        private DbSet<T> DbSet { get; }
+
+        /// <inheritdoc/>
+        public Task AddEntityAsync(T entity)
+        {
+            var validatedEntity = entity ?? throw new ArgumentNullException(nameof(entity));
+
+            DbSet.AddAsync(validatedEntity);
+
+            return SaveAsync();
+        }
+
+        /// <inheritdoc/>
+        public Task<List<T>> AllAsync() => this.DbSet.Where(x => !x.IsDeleted).ToListAsync();

```

Comment:
Consider to make this virtual.
Use IEnumerable<T> as a return type or IReadOnlyList<T>

Reply:
In this case I can not set `Task<IEnumerable<T>>`, but I can
`public IAsyncEnumerable<T> AllAsync()
 => this.DbSet.Where(x => !x.IsDeleted).ToAsyncEnumerable();`, but in this case we should consider using `IAsyncEnumerable` everywhere.

Reply:
Use IQueryable then.